tscv/_split.py

```diff
@@ -371,6 +372,128 @@ def get_n_splits(self, X=None, y=None, groups=None):
         return self.n_splits
 
 
+class CombinatorialGapKFold(GapCrossValidator):
+    """Combinatorial K-Folds cross-validator with Gaps
+
+    Provides train/test indices to split data in train/test sets. Split
+    dataset into N groups of k folds (without shuffling).
+
+    Parameters
+    ----------
+    N : int, default=5
+        Number of groups. Must be at least 2.
+
+    k : int, default=2
+        Number of test splits. Must be at least 1.
+
+    gap_before : int, default=0
+        Gap before the test sets.
+
+    gap_after : int, default=0
+        Gap after the test sets.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from tscv import CombinatorialGapKFold
+    >>> cgkf = CombinatorialGapKFold(N=5, k=2, gap_before=1, gap_after=1)
+    >>> cgkf.get_n_splits(np.arange(10))
+    10
+    >>> print(cgkf)
+    CombinatorialGapKFold(N=None, gap_after=1, gap_before=1, k=None)
+    >>> for train_index, test_index in cgkf.split(np.arange(10)):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    TRAIN: [5 6 7 8 9] TEST: [0 1 2 3]
+    TRAIN: [7 8 9] TEST: [0 1 4 5]
+    TRAIN: [3 4 9] TEST: [0 1 6 7]
+    TRAIN: [3 4 5 6] TEST: [0 1 8 9]
+    TRAIN: [0 7 8 9] TEST: [2 3 4 5]
+    TRAIN: [0 9] TEST: [2 3 6 7]
+    TRAIN: [0 5 6] TEST: [2 3 8 9]
+    TRAIN: [0 1 2 9] TEST: [4 5 6 7]
+    TRAIN: [0 1 2] TEST: [4 5 8 9]
+    TRAIN: [0 1 2 3 4] TEST: [6 7 8 9]
+    """
+
+    def __init__(self, N=5, k=2, gap_before=0, gap_after=0):
+        if not isinstance(N, numbers.Integral):
+            raise ValueError('The number of groups must be of Integral type. '
+                             '%s of type %s was passed.'
+                             % (N, type(N)))
+        N = int(N)
+
+        if not isinstance(k, numbers.Integral):
+            raise ValueError('The number of test splits must be of Integral '
+                             'type. %s of type %s was passed.'
+                             % (k, type(k)))
+        k = int(k)
+
+        if N <= 1:
+            raise ValueError(
+                "Combinatorial k-fold cross-validation requires at least two"
+                " groups by setting N=2 or more,"
+                " got N={0}.".format(N))
+
+        if k < 1:
+            raise ValueError(
+                "Combinatorial k-fold cross-validation requires at least one"
+                " test split by setting k=1 or more,"
+                " got k={0}.".format(k))
+
+        super().__init__(gap_before, gap_after)
+        self.n_groups = N
+        self.test_splits = k
+
+    def split(self, X, y=None, groups=None):
```

Comment:
The canonical way of doing this is to redefine `_iter_test_indices(self, X, y=None, groups=None)` from the base class and generate only the test indices. The base class will take care of the rest. Please refer to the other derived classes and make modification accordingly.

Reply:
The check of the training set size can be delayed and implemented in `GapCrossValidator._iter_train_indices()` and its 3 siblings. You can also implement it in `GapCrossValidator.split()` if you find it a hustle to implement it four times.

Reply:
On second thought, I don't find it necessary to check the non-emptiness of the training set. Some models/algorithms can output a default estimator/strategy given an empty training set (e.g., equal weight portfolio). If your model/algorithm requires the non-empty training set, it's probably a good idea to check it in the model/algorithm rather than in the cross-validator. That's why I didn't implement this check when I released the package. 

That said, I probably forgot to check the non-emptiness of the test set.

Reply:
Ok then, I will replace the implementation of the split method with _iter_test_indices

About the non-emptiness of the test set I think you prefer to implement it in the base class with a specific PR, so I won't touch it, ok?

Reply:
Your judgement is correct.